### PR TITLE
MAILBOX-374 Add retry mechanism for flag updates

### PR DESCRIPTION
**Pull request by blackheaven**

### CODE REVIEW

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/CassandraAsyncExecutor.java [hunk 1]
```diff
@@ -76,4 +80,23 @@ public CassandraAsyncExecutor(Session session) {
         return executeSingleRow(statement)
                 .hasElement();
     }
+
+    private Mono<ResultSet> retryOnError(Mono<ResultSet> query) {
+        return query
+                .retryWhen(companion -> companion
```
Comment (trantienduchn) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/CassandraAsyncExecutor.java]: why not the retry backoff ? :)

Comment (blackheaven) [reply backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/CassandraAsyncExecutor.java]: It's not few lines after:

```
                 .flatMap(index -> Mono.delay(Duration.ofMillis(index * 100)))
```

You have to filter on the `Exception` class, that's why I've used a more general operator.

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/CassandraAsyncExecutor.java [hunk 2]
```diff
@@ -42,9 +46,9 @@ public CassandraAsyncExecutor(Session session) {
     }
 
     public Mono<ResultSet> execute(Statement statement) {
-        return Mono.defer(() -> Mono.fromFuture(FutureConverter
+        return retryOnError(Mono.defer(() -> Mono.fromFuture(FutureConverter
```
Comment (rouazana) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/CassandraAsyncExecutor.java]: how do you know you can retry? Some operations are idempotent, but some other are not, no? For example increasing quota count.

Comment (blackheaven) [reply backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/CassandraAsyncExecutor.java]: Sure, but you do not know that it has been run or not.

Comment (rouazana) [reply backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/CassandraAsyncExecutor.java]: Yes, that's my point. Retry should not be for all operations, just for operations that you know they can be retried.

**Comment by chibenwa** [comment]
Why do we need this? Maybe you are trying a too general solution for maybe very specific problem.

Let's investigate before taking out the bazooka...

(By over-retrying on cassandra, we might hammer even more a freshly rebooted Cassandra for instance, getting the problem worse, and over stressing maybe surviving nodes. In some cases failing requests is better than getting the system down)

**Comment by blackheaven** [comment]
I think your comment apply to the first version of this PR

**Comment by chibenwa** [comment]
The reworked version still applies now

**Comment by blackheaven** [comment]
So, not really, because a new Cassandra instance has, by definition, no data, so no retry will be done.

In the normal case, only flag updates and swift blob reading are replayed, it does not seem much.

Regarding the need, at some point, Cassandra "forget to answer", either we ask the user to resubmit its mail, either we try again, it's not really glorious, but I do not have enough master of Cassandra driver to strenghten the connection, which would be the ideal way to do so.

**Comment by chibenwa** [comment]
https://docs.datastax.com/en/drivers/java/2.0/com/datastax/driver/core/policies/RetryPolicy.html

 + set RetryPolicy ?

Also, did you reproduced this behaviour? With a test?

**Comment by blackheaven** [comment]
The problem does not change whether it is done at Cassandra level or at Reactor level, you have to replay the query, and you will "overload" Cassandra.

Sure, I'm manually testing it, I do not reproduce our problem with these fixes.

**Comment by chibenwa** [comment]
> The problem does not change whether it is done at Cassandra level or at Reactor level, you have to replay the query, and you will "overload" Cassandra.

true

**Comment by chibenwa** [comment]
We may find mitigation strategies, where we store stuff in alternative data-stores (files?) so that it could get retried later on.

I agree this is much work compared to the simple retry solution you propose. Agree with the retry then...

A test however would really be cool IMO

**Comment by blackheaven** [comment]
O.o it would involve to query multiple data-stores, unifying things, and so on, it looks very complicated.

**Comment by chibenwa** [comment]
> O.o it would involve to query multiple data-stores, unifying things, and so on, it looks very complicated.

Yes so let's not do it, and do retry.

But please consider using `RetryPolicies` that allows more control on the retry (like downgrading consistency) and is thus more evolutive.

Also I would REALLY like to have a test on this. Could you try to break things with a concurrent test?

**Comment by Arsnael** [comment]
Read it...

**Comment by trantienduchn** [comment]
> Read it...

me too

**Comment by blackheaven** [comment]
I have added a test but it does not break it :/

**Comment by blackheaven** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by hoangdat** [comment]
I read it but not confident for a approval

**Comment by rouazana** [comment]
I agree that we should at least remove the associated configuration option

**Comment by blackheaven** [comment]
test this please

**Comment by Arsnael** [comment]
test this please

**Comment by blackheaven** [comment]
test this please

**Comment by blackheaven** [comment]
test this please

**Comment by blackheaven** [comment]
Finally :green_apple:

File: server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java [hunk 3]
```diff
@@ -119,20 +121,24 @@
 
     @Override
     public InputStream read(BlobId blobId) throws ObjectStoreException {
-        Blob blob = blobStore.getBlob(containerName.value(), blobId.asString());
+        Optional<Blob> retrievedBlob = Mono.fromCallable(() -> Optional.ofNullable(blobStore.getBlob(containerName.value(), blobId.asString())))
+            .retryBackoff(5, Duration.ofMillis(100))
```
Comment (mbaechler) [inline server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java]: magic numbers ?

Comment (blackheaven) [reply server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java]: Do you want some configuration or Constants are enough?

Comment (mbaechler) [reply server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java]: constants are ok.

**Comment by blackheaven** [comment]
Splitted in #2236 and #2237

**Pull request closed without merging**